### DISCUSSION TITLE
Question: How to filter optimiser results based on some boolean check?

### DISCUSSION THREAD

**Discussion by TheCodeSummoner** [do-mpc/do-mpc#128] [Q&A] [answered]
Hi,

I am once again in need of advice and kindly ask for your help.

Thanks to Your brilliant tool I have managed to implement trajectory following for a 4-wheeled vehicle. For simplicity, consider the following distance-based cost function:

```python
def terminal_cost(self):
    return (self._target_x - self._position_x) ** 2 + (self._target_y - self._position_y) ** 2
```

This allows the vehicle to follow some chosen coordinate.

However, the most optimal path chosen by this cost function may pass through obstacles and make it impossible for the car to physically reach the target point. Therefore, I wanted to _exclude_ all paths that are outside of the track or passing through an obstacle. Thus, I want to either completely discard unsafe trajectories or introduce a binary-like stage cost function.

Let's say that a safe trajectory is defined as a collection of points in which **every** point passes the following check:

```python
from shapely.geometry import Polygon, Point
Polygon(points).contains(Point(position_x, position_y))
```

In other words, every position within the horizon must be inside some polygon (safe region).

Is there a way to either mark this as a check that each prediction must meet, or define this as a stage cost function? Something like:

```python
def _some_casadi_wrapper(self):
    # self._points can be _tvp, whereas self._position_x and self._position_y will be _x
    return 0 if Polygon(self._points).contains(Point(self._position_x, self._position_y)) else 1 

def stage_cost(self):
    # 10000 is a big weight to make sure the unsafe trajectories are discarded
    # Notice that the result of this is either 0 or 10000 (really high value)
    return 10000 * self._some_casadi_wrapper
```

Thank you for your help!

**Comment by 4flixt** [comment, marked as answer]
Hi CodeSummoner,

I converted this issue to a discussion. Also thank you for your question and the endorsement. 

If you can formulate your safe region as some sort of non-linear constraint you can use the [set_nl_cons](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.set_nl_cons.html) method of the ``MPC`` class. 
This would definitely be the preferred method. You need to implement these constraints in terms of CasADi variables. 

I cannot recommend your proposed method which is similar to an indicator function. Typically that would lead to numerically very challenging problems.
I would also refrain from using casadi methods such as if_else (they exists and should work in principle) for the same reason.

Also note that **do-mpc** cannot solve mixed integer problems with binary decision variables. These are much harder problems.

I hope this gives you some ideas :)

Best
Felix

**Reply by TheCodeSummoner** [reply]
Thank you for your answer and apologies for posting an issue (I was not aware of the _discussion_ section).

I will have a look at your recommendations.

Kind regards,
Kacper

**Reply by 4flixt** [reply]
All good the Discussions tab has just recently been added to Github :)

**Reply by TheCodeSummoner** [reply]
I've been progressing a little bit with `set_nl_cons`. Apologies if you have been notified about the (already deleted) answer, which I have removed since I just realised that the upper bound is a constant and I can't change it dynamically.

Do you reckon there is a better way to achieve the following?

```python
self._controller.set_nl_cons("some-name", self._a * self._y + self._b * self._x + self._c, 0)
```

Since the safety region (polygon) will be moving with time, it must be modelled using time-varying parameters. Therefore, **every** constraint will be in the form of `ay + bx + c <= 0`. If I approximate the polygon using 5 inequalities, that gives a grand total of `5*3=15`  time-varying parameters (!) (plus additional 2 parameters for modelling the position in the above example). I'm slightly worried this will be an overkill for the tool, especially if I decide to have more inequalities.

Is there maybe some vector form I could use? As in `Ay + By + C <= 0`, where each row in A, B, and C is used to form the above inequality `ay + bx + c <= 0`, and then the two variables (vectors) will be the time-varying parameters.

**Reply by 4flixt** [reply]
Hi @TheCodeSummoner,

I think your approach looks reasonable. Parameterizing a polgon just takes a decent amount of parameters. There is no way around it except for simplifying the shape. Vectorizing everything will probably simplify your notation but otherwise the tool doesn't profit from it in this case. 

In any case it shouldn't be a major issue to have a large amount of time-varying-parameters. They aren't optimization variables so it won't slow down the process significantly. 
The problem is rather that every constraint adds complexity to the problem (additional lagrange multipliers for example). So I would try to keep the shapes as simple as possible. 

Best
4flixt
